Missing environment information could be better diagnosed

I recently got this:

```
Errors for silver:extension:testing
 [/home/tedinski/repos/silver/grammars/silver/extension/testing/EqualityTest.sv]
EqualityTest.sv:67:95: error: Attribute 'grammarFlowTypes' does not occur on 'Decorated silver:driver:util:RootSpec'. Looked at:
	silver:definition:env:grammarFlowTypes (Attributes.sv:82)
EqualityTest.sv:68:103: error: Attribute 'productionFlowGraphs' does not occur on 'Decorated silver:driver:util:RootSpec'. Looked at:
	silver:definition:env:productionFlowGraphs (Attributes.sv:81)
```

Technically, this error is correct, but the root cause was `silver:driver:util` not being imported, which is where `RootSpec` and its occurrences are declared.

Since `nonterminal Foo with ...` is a common case, I think if we run into an error involving a type from a grammar that's not imported, perhaps we should point that out as part of the error message.

--- a/grammars/silver/definition/core/Expr.sv
+++ b/grammars/silver/definition/core/Expr.sv
@@ -1057,20 +1057,29 @@ Boolean ::= l::Pair<Integer Decorated Expr>  r::Pair<Integer Decorated Expr> { r
 function mkStrFunctionInvocation
 Expr ::= l::Location  e::String  es::[Expr]
 {
-  return mkFunctionInvocation(l, baseExpr(qName(l, e), location=l), es);
+  return mkFullFunctionInvocation(l, baseExpr(qName(l, e), location=l), es, []);
 }
 function mkFunctionInvocation
 Expr ::= l::Location  e::Expr  es::[Expr]
 {
-  return application(e, '(', foldAppExprs(l, reverse(es)), ',', emptyAnnoAppExprs(location=l), ')', location=l);
+  return mkFullFunctionInvocation(l, e, es, []);
 }
--- WARNING: NOTE THAT YOU NEED TO REVERSE THE EXPR LIST BEFORE CALLING THIS:
-function foldAppExprs
-AppExprs ::= l::Location  e::[Expr]
+function mkFullFunctionInvocation
+Expr ::= l::Location  e::Expr  es::[Expr]  ans::[Pair<String Expr>]
 {
-  return if null(e) then emptyAppExprs(location=l)
-         else if null(tail(e)) then oneAppExprs(presentAppExpr(head(e), location=l), location=l)
-         else snocAppExprs(foldAppExprs(l, tail(e)), ',', presentAppExpr(head(e), location=l), location=l);
+  return application(e, '(',
+    foldl(snocAppExprs(_, ',', _, location=l), emptyAppExprs(location=l),
+      map(presentAppExpr(_, location=l), es)),
+    ',',
+    foldl(snocAnnoAppExprs(_, ',', _, location=l), emptyAnnoAppExprs(location=l),
+      map(mkAnnoExpr, ans)),
+    ')', location=l);
+}
+-- Internal helper function
+function mkAnnoExpr
+AnnoExpr ::= p::Pair<String Expr>
+{
+  return annoExpr(qName(p.snd.location, p.fst), '=', presentAppExpr(p.snd, location=p.snd.location), location=p.snd.location);
 }
 
 {--

--- a/grammars/silver/definition/core/QName.sv
+++ b/grammars/silver/definition/core/QName.sv
@@ -146,14 +146,19 @@ synthesized attribute attrDcl :: DclInfo;
  -}
 synthesized attribute found :: Boolean;
 
+{--
+ - Used like `x.<this>`.
+ - @param  at       the name of an attribute
+ - @inh    attrFor  the type this attribute should be on
+ -}
 concrete production qNameAttrOccur
 top::QNameAttrOccur ::= at::QName
 {
   top.name = at.name;
   top.unparse = at.unparse;
   
-  -- Start with `at.lookupAttribute.dcls`
-  -- Occurrence lookups for each attribute dcl on `top.attrFor`
+  -- We start with all attributes we find with the name `at`:
+  -- Then we filter to just those that appear to have an occurrence on `top.attrFor`:
   local narrowed :: [[DclInfo]] = 
     -- The occurs dcls on this nonterminal for
     map(getOccursDcl(_, top.attrFor.typeName, top.env),
@@ -176,20 +181,26 @@ top::QNameAttrOccur ::= at::QName
       length(attrsNarrowed) != 1);
   
   top.errors :=
+    -- If we fail to look up the attribute, just report that.
     if null(at.lookupAttribute.dcls) then
       at.lookupAttribute.errors
+    -- If we're looking up an attribute on `errorType`, an error is already raised, don't create noise
     else if top.attrFor.isError then
-      -- suppress further errors
       []
-    else if null(dclsNarrowed) then 
-      -- Note we're using the short name of the attribute... there may be more than one attribute Dcl
-      -- But, none of them occur so this error quite suffices!
-      [err(at.location, "Attribute '" ++ at.name ++ "' does not occur on '" ++ prettyType(top.attrFor) ++ "'. Looked at:\n" ++ printPossibilities(at.lookupAttribute.dcls))]
+    -- If no attribute occurs on this type, raise that error
+    else if null(dclsNarrowed) then
+      -- This is a heuristic error message for the situation where you have a type, but haven't imported
+      -- the grammar declaring that type.
+      (if top.attrFor.typeName != "" && null(getTypeDcl(top.attrFor.typeName, top.env)) then
+         [err(at.location, "Attribute '" ++ at.name ++ "' does not occur on '" ++ prettyType(top.attrFor) ++ "'. Perhaps import '" ++ substring(0, lastIndexOf(":", top.attrFor.typeName), top.attrFor.typeName)  ++ "'?")]
+       else
+         [err(at.location, "Attribute '" ++ at.name ++ "' does not occur on '" ++ prettyType(top.attrFor) ++ "'. Looked at:\n" ++ printPossibilities(at.lookupAttribute.dcls))]
+      )
+    -- If more than one attribute on the same _short name_ occurs, raise ambiguity
     else if length(attrsNarrowed) > 1 then
-      -- Here we've found multiple attributes that occur here!
       [err(at.location, "Ambiguous reference to attribute occurring on '" ++ at.name ++ "'. Possibilities are:\n" ++ printPossibilities(attrsNarrowed))]
+    -- If this same attribute has multiple occurences (must be due to orphaned occurs)
     else []; {-if length(dclsNarrowed) > 1 then
-      -- This should be essentially impossible! (Requires "orphaned occurs" which are currently allowed...)
       [err(at.location, "There are erroneously multiple attribute occurrences for '" ++ at.name ++ "'. Possibilities are:\n" ++ printPossibilities(dclsNarrowed))]
     else [];-}
     -- TODO: This last bit is disabled because we have problems with importing grammars multiple times.

--- a/grammars/silver/extension/auto_ast/AutoAst.sv
+++ b/grammars/silver/extension/auto_ast/AutoAst.sv
@@ -59,30 +59,18 @@ top::ProductionStmt ::= 'abstract' v::QName ';'
       '.',
       astQName,
       '=',
-      application(
+      mkFullFunctionInvocation(
+        top.location,
         baseExpr(v, location=v.location),
-        '(',
-        -- just a note here that 'foldAppExprs' needs the exprs reversed...
-        -- sort of using an internal function here.
-        foldAppExprs(top.location, reverse(map(accessAst(_, top.location), elems))),
-        ',',
-        if hasLoc
-        then oneAnnoAppExprs(
-          annoExpr(
-            qName(top.location, "location"),
-            '=',
-            presentAppExpr(
-              access(
-                baseExpr(lhsQName, location=top.location),
-                '.',
-                qNameAttrOccur(qName(top.location, "location"), location=top.location),
-                location=top.location),
-              location=top.location),
-            location=top.location),
-          location=top.location)
-        else emptyAnnoAppExprs(location=top.location),
-        ')',
-        location=top.location),
+        map(accessAst(_, top.location), elems),
+        if hasLoc then
+         [pair("location", 
+            access(
+              baseExpr(lhsQName, location=top.location),
+              '.',
+              qNameAttrOccur(qName(top.location, "location"), location=top.location),
+              location=top.location))]
+        else []),
       ';',
       location=top.location);
 }

--- a/grammars/silver/extension/functorattrib/ProductionBody.sv
+++ b/grammars/silver/extension/functorattrib/ProductionBody.sv
@@ -12,7 +12,7 @@ top::ProductionStmt ::= 'propagate' ns::NameList ';'
   -- and propagateAttrDcl containing the remaining names
   forwards to
     case ns of
-      nameListOne(n) -> 
+    | nameListOne(n) -> 
         propagateOne(n, location=top.location)
     | nameListCons(n, _, rest) ->
         productionStmtAppend(
@@ -23,84 +23,68 @@ top::ProductionStmt ::= 'propagate' ns::NameList ';'
 }
 
 {--
- - Generates the list of AppExprs used in calling the constructor
+ - Generates the expression we should use for an argument
  - @param loc      The parent location to use in construction
  - @param env      The environment
  - @param attrName The name of the attribute being propagated
- - @param inputs   The NamedSignatureElements being propagated
- - @return A list of AppExprs to be used to build the arguments
+ - @param input    The NamedSignatureElement being propagated
+ - @return Either this the child, or accessing `attrName` on the child
  -}
-function makeArgs
-[AppExpr] ::= loc::Location env::Decorated Env attrName::QName inputs::[NamedSignatureElement]
+function makeArg
+Expr ::= loc::Location env::Decorated Env attrName::QName input::NamedSignatureElement
 {
   attrName.env = env;
   
-  local at::QName = qName(loc, head(inputs).elementName);
+  local at::QName = qName(loc, input.elementName);
   at.env = env;
   
   -- Check if the attribute occurs on the first child
   local attrOccursOnHead :: Boolean = 
     !null(
-        -- The occurs dcls on this nonterminal for
-      flatMap(getOccursDcl(_, head(inputs).typerep.typeName, env),
+      -- The occurs dcls on this nonterminal for
+      flatMap(getOccursDcl(_, input.typerep.typeName, env),
           -- the full names of each candidate
           map((.fullName), attrName.lookupAttribute.dcls)));
-  local validTypeHead :: Boolean = head(inputs).typerep.isDecorable;
+  local validTypeHead :: Boolean = input.typerep.isDecorable;
   
   return
-    case inputs of
-      [] -> []
-    | h :: rest -> 
-        (if validTypeHead && attrOccursOnHead
-         then presentAppExpr(
-                access(
-                  baseExpr(at, location=loc), '.',
-                  qNameAttrOccur(attrName, location=loc),
-                  location=loc),
-                location=loc)
-         else presentAppExpr(
-                baseExpr(at, location=loc),
-                location=loc)) :: makeArgs(loc, env, attrName, rest)
-    end;
+    if validTypeHead && attrOccursOnHead
+    then access(
+           baseExpr(at, location=loc), '.',
+           qNameAttrOccur(attrName, location=loc),
+           location=loc)
+    else baseExpr(at, location=loc);
 }
 
 {--
  - Generates the list of AnnoExprs used in calling the constructor
  - @param loc      The parent location to use in construction
  - @param baseName The name of the parent from the signature
- - @param inputs   The NamedSignatureElements for the annotations
+ - @param input   The NamedSignatureElement for an annotation
  - @return A list of AnnoExprs to be used to build the named arguments
  -}
-function makeAnnoArgs
-[AnnoExpr] ::= loc::Location baseName::QName inputs::[NamedSignatureElement]
+function makeAnnoArg
+Pair<String Expr> ::= loc::Location baseName::QName input::NamedSignatureElement
 {
   -- TODO: This is a hacky way of getting the base name, not sure if correct
-  local annoName::QName = qName(loc, last(explode(":", head(inputs).elementName)));
+  -- trouble is the annotations are listed as fullnames, but have to be supplied as shortnames. weird.
+  local annoName :: String = last(explode(":", input.elementName));
 
   return
-    case inputs of
-      [] -> []
-    | h :: rest -> 
-      annoExpr(
-        annoName,
-        '=',
-        presentAppExpr(
-          access(
-            baseExpr(baseName, location=loc), '.',
-              qNameAttrOccur(annoName, location=loc),
-              location=loc),
-            location=loc),
-          location=loc) :: makeAnnoArgs(loc, baseName, rest)
-    end;
+    pair(annoName,
+      access(
+        baseExpr(baseName, location=loc), '.',
+        qNameAttrOccur(qName(loc, annoName), location=loc),
+        location=loc));
 }
 
 -- In the future, this should maybe be dispatch for different types of attributes (e.g. monoid)
 {--
  - Propagate a functor attribute on the enclosing production
- - @param a  The name of the attribute to propagate
+ - @param attr  The name of the attribute to propagate
  -}
 abstract production propagateOne
-top::ProductionStmt ::= a::QName
+top::ProductionStmt ::= attr::QName
 {
   -- No explicit errors, for now.  The only conceivable issue is the attribute not
   -- occuring on the LHS but this should be caught by the forward errors.  
@@ -111,26 +95,24 @@ top::ProductionStmt ::= a::QName
   prodName.grammarName = top.grammarName;
   prodName.config = top.config;
   prodName.env = top.env;
-  local inputs::AppExprs = 
-    foldl(snocAppExprs(_, ',', _, location=top.location),
-          emptyAppExprs(location=top.location),
-          makeArgs(top.location, top.env, a, top.frame.signature.inputElements));
-  local annotations::AnnoAppExprs = 
-    foldl(snocAnnoAppExprs(_, ',', _, location=top.location),
-          emptyAnnoAppExprs(location=top.location),
-          makeAnnoArgs(top.location, topName, top.frame.signature.namedInputElements));
+
+  local inputs :: [Expr] = 
+    map(makeArg(top.location, top.env, attr, _), top.frame.signature.inputElements);
+  local annotations :: [Pair<String Expr>] = 
+    map(makeAnnoArg(top.location, topName, _), top.frame.signature.namedInputElements);
 
   -- Construct an attribute def and call with the generated arguments
   forwards to 
     attributeDef(
       concreteDefLHS(topName, location=top.location),
       '.',
-      qNameAttrOccur(a, location=top.location),
+      qNameAttrOccur(attr, location=top.location),
       '=',
-      application(
+      mkFullFunctionInvocation(
+        top.location,
         baseExpr(prodName, location=top.location),
-        '(', inputs, ',', annotations, ')',
-        location=top.location),
+        inputs,
+        annotations),
       ';',
       location=top.location);
 }

--- a/grammars/silver/extension/monad/AbstractSyntax.sv
+++ b/grammars/silver/extension/monad/AbstractSyntax.sv
@@ -6,29 +6,18 @@ top::Expr ::= n::Name t::TypeExpr e::Expr rest::Expr bindFn::QName
   rest.downSubst = e.upSubst;
   forward.downSubst = rest.upSubst;-}
 
-  forwards to
-    applicationExpr(
-      baseExpr(bindFn, location=top.location),
-      '(',
-      snocAppExprs(
-        oneAppExprs(
-          presentAppExpr(e, location=top.location),
-          location=top.location),
-        ',',
-        presentAppExpr(
-          lambdap(
-            productionRHSCons(
-              productionRHSElem(
-                n, '::', t,
-                location=top.location),
-              productionRHSNil(location=top.location),
-              location=top.location),
-            rest,
-            location=top.location),
-          location=top.location),
+  local cont :: Expr =
+    lambdap(
+      productionRHSCons(
+        productionRHSElem(
+          n, '::', t, location=top.location),
+        productionRHSNil(location=top.location),
         location=top.location),
-      ')',
+      rest,
       location=top.location);
+
+  forwards to
+    mkFunctionInvocation(top.location, baseExpr(bindFn, location=top.location), [e, cont]);
 }
 
 abstract production returnExpr
@@ -41,11 +30,7 @@ top::Expr ::= e::Expr isFinalVal::Boolean returnFn::QName
     else [];
 
   forwards to
-    applicationExpr(
-      baseExpr(returnFn, location=top.location), '(',
-      oneAppExprs(presentAppExpr(e, location=top.location), location=top.location),
-      ')',
-      location=top.location);
+    mkFunctionInvocation(top.location, baseExpr(returnFn, location=top.location), [e]);
 }
 
 abstract production returnUnitExpr
@@ -53,8 +38,6 @@ top::Expr ::= returnFn::QName
 {
   forwards to
     returnExpr(
-      applicationEmpty(
-        baseExpr(qName(top.location, "unit"), location=top.location), '(', ')',
-        location=top.location),
+      mkStrFunctionInvocation(top.location, "core:unit", []),
       true, returnFn, location=top.location);
 }

--- a/grammars/silver/extension/silverconstruction/Translation.sv
+++ b/grammars/silver/extension/silverconstruction/Translation.sv
@@ -20,13 +20,11 @@ top::AST ::= prodName::String children::ASTs annotations::NamedASTs
         case reify(a) of
         | right(e) ->
           just(
-            applicationExpr(
-              baseExpr(
-                makeQName("silver:extension:silverconstruction:makeQName", givenLocation),
-                location=givenLocation),
-              '(',
-              foldAppExprs(givenLocation, [e, locAST.translation]),
-              ')', location=givenLocation))
+            mkFullFunctionInvocation(
+              givenLocation,
+              baseExpr(qName(givenLocation, "silver:hostEmbedding:makeQName"), location=givenLocation),
+              [e, locAST.translation],
+              []))
         | left(msg) -> error(s"Error in reifying child of production ${prodName}:\n${msg}")
         end
     | "silver:extension:silverconstruction:escape_qName", _, _ ->
@@ -36,13 +34,11 @@ top::AST ::= prodName::String children::ASTs annotations::NamedASTs
         case reify(a) of
         | right(e) ->
           just(
-            applicationExpr(
-              baseExpr(
-                makeQName("silver:extension:silverconstruction:makeName", givenLocation),
-                location=givenLocation),
-              '(',
-              foldAppExprs(givenLocation, [e, locAST.translation]),
-              ')', location=givenLocation))
+            mkFullFunctionInvocation(
+              givenLocation,
+              baseExpr(qName(givenLocation, "silver:hostEmbedding:makeName"), location=givenLocation),
+              [e, locAST.translation],
+              []))
         | left(msg) -> error(s"Error in reifying child of production ${prodName}:\n${msg}")
         end
     | "silver:extension:silverconstruction:escape_name", _, _ ->

--- a/grammars/silver/extension/treegen/Arbitrary.sv
+++ b/grammars/silver/extension/treegen/Arbitrary.sv
@@ -116,16 +116,19 @@ Expr ::= index::Integer  lst::[DclInfo]  total::Integer  l::Location
 function deriveGenerateOn
 Expr ::= id::DclInfo  l::Location
 {
-  local callid :: Expr = baseExpr(qName(l, id.fullName), location=l);
-  local es :: [Expr] = map(callGenArb(_, l), id.typerep.inputTypes);
-  local annos :: AnnoAppExprs =
+  local annos :: [Pair<String Expr>] =
     if null(id.typerep.namedTypes) then
-      emptyAnnoAppExprs(location=l)
+      []
     else
-      -- erroneously assume it's location for now. just supply location=bogusLoc()
-      oneAnnoAppExprs(annoExpr(qName(l, "location"), '=', presentAppExpr(mkStrFunctionInvocation(l, "bogusLoc", []), location=l), location=l),location=l);
+      -- we just erroneously assume the annotation must be location, for now
+      [pair("location", mkStrFunctionInvocation(l, "bogusLoc", []))];
+
   return
-    application(callid, '(', foldAppExprs(l, reverse(es)), ',', annos, ')', location=l);
+    mkFullFunctionInvocation(
+      l,
+      baseExpr(qName(l, id.fullName), location=l),
+      map(callGenArb(_, l), id.typerep.inputTypes),
+      annos);
 }
 
 -- Call generateArbitraryID

--- a/grammars/silver/hostEmbedding/Translation.sv
+++ b/grammars/silver/hostEmbedding/Translation.sv
@@ -97,16 +97,11 @@ top::AST ::= prodName::String children::ASTs annotations::NamedASTs
     
   top.translation =
     fromMaybe(
-      application(
-        baseExpr(makeQName(prodName, givenLocation), location=givenLocation),
-        '(',
-        foldAppExprs(givenLocation, reverse(children.translation)),
-        ',',
-        foldl(
-          snocAnnoAppExprs(_, ',', _, location=givenLocation),
-          emptyAnnoAppExprs(location=givenLocation),
-          reverse(annotations.translation)),
-        ')', location=givenLocation),
+      mkFullFunctionInvocation(
+        givenLocation,
+        baseExpr(qName(givenLocation, prodName), location=givenLocation),
+        children.translation,
+        annotations.translation),
       escapeTranslation);
   
   children.givenLocation = givenLocation;
@@ -215,7 +210,7 @@ top::ASTs ::=
   top.foundLocation = nothing();
 }
 
-attribute givenLocation, translation<[AnnoExpr]>, foundLocation occurs on NamedASTs;
+attribute givenLocation, translation<[Pair<String Expr>]>, foundLocation occurs on NamedASTs;
 
 aspect production consNamedAST
 top::NamedASTs ::= h::NamedAST t::NamedASTs
@@ -231,17 +226,14 @@ top::NamedASTs ::=
   top.foundLocation = nothing();
 }
 
-attribute givenLocation, translation<AnnoExpr>, foundLocation occurs on NamedAST;
+attribute givenLocation, translation<Pair<String Expr>>, foundLocation occurs on NamedAST;
 
 aspect production namedAST
 top::NamedAST ::= n::String v::AST
 {
   top.translation =
-    annoExpr(
-      qNameId(makeName(last(explode(":", n)), top.givenLocation), location=top.givenLocation),
-      '=',
-      presentAppExpr(v.translation, location=top.givenLocation),
-      location=top.givenLocation);
+    -- hack to get annotation shortname
+    pair(last(explode(":", n)), v.translation);
   top.foundLocation =
     if n == "core:location"
     then
@@ -251,6 +243,9 @@ top::NamedAST ::= n::String v::AST
       end
     else nothing();
 }
+
+-- the functions below are directly referenced in reflection code in silver:extensions:silverconstruction
+-- so make sure you grep for that if you change/move them.
 
 function makeName
 Name ::= n::String loc::Location

